contracts/src/Tracker.sol

```diff
@@ -97,17 +110,37 @@ contract Tracker is EIP712 {
         emit Join(msg.sender, challengeId);
     }
 
-    // todo: change signature to bytes, so we can support contract checkins
-    // todo: change checkin digest into bytes, to be more flexible
-    function checkIn(uint256 challengeId, uint256 timestamp, uint8 v, bytes32 r, bytes32 s) public {
+    /**
+     * challengeId is required to prevent replays from other challenges in case the same verifiers are used.
+     * checkInData is application specific, what to check on and replay prevention depend on the implementation of the checkInJudge contract specified at challenge creation.
+     * If the checkInJudge contract is specified, checkInData would be pass with the call checkInJudge.judge(checkInData).
+     * signature is generated with signing the returned digest from getCheckInDigest(uint256 challengeId, address user, bytes checkInData);
```

Comment:
the comment here is a bit confusing, kind of implying the contract is generating the signature.
```suggestion
     * @signature checkInDigest signed by verifier address;
```